### PR TITLE
Encryption Profile

### PR DESCRIPTION
**Pull request by nerdstein**
1. Adds new config entity for "Encryption Profile"
2. Functionality for this entity to have "default" to make the service return a default
3. Updates service to allow an instance to be passed

### CODE REVIEW

**Comment by nerdstein** [comment]
@doylejd , @damontgomery - feedback welcomed

**Comment by nerdstein** [comment]
Known gaps - need to update the automated tests

File: src/Entity/EncryptionConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * Contains Drupal\encrypt\Entity\EncryptionConfiguration.
+ */
+
+namespace Drupal\encrypt\Entity;
+
+use Drupal\Core\Config\Entity\ConfigEntityBase;
+use Drupal\encrypt\EncryptionConfigurationInterface;
+
+/**
+ * Defines the Title entity.
```
Comment (doylejd) [inline src/Entity/EncryptionConfiguration.php]: This should be "Encryption" entity not title.

File: src/Controller/EncryptionConfigurationListBuilder.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * Contains Drupal\encrypt\Controller\EncryptionConfigurationListBuilder.
+ */
+
+namespace Drupal\encrypt\Controller;
+
+use Drupal\Core\Config\Entity\ConfigEntityListBuilder;
+use Drupal\Core\Entity\EntityInterface;
+
+/**
+ * Provides a listing of Title entities.
```
Comment (doylejd) [inline src/Controller/EncryptionConfigurationListBuilder.php]: This should be "Encryption" entity not title.

File: src/EncryptionConfigurationInterface.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * @file
+ * Contains Drupal\encrypt\EncryptionConfigurationInterface.
+ */
+
+namespace Drupal\encrypt;
+
+use Drupal\Core\Config\Entity\ConfigEntityInterface;
+
+/**
+ * Provides an interface for defining Title entities.
```
Comment (doylejd) [inline src/EncryptionConfigurationInterface.php]: This should be "Encryption" entity not title.

**Comment by doylejd** [comment]
@nerdstein  Do we need to define a schema for the new entity type?

File: src/EncryptService.php [hunk 4]
```diff
@@ -18,50 +18,63 @@
...
+  /**
    * @var \Drupal\encrypt\EncryptionMethodManager
    */
-  protected $manager;
+  protected $encryptManager;
 
   /**
    * @var \Drupal\key\KeyManager
    */
   protected $key;
```
Comment (doylejd) [inline src/EncryptService.php]: To keep things consistent we should rename this to $keyManager.

File: src/Form/EncryptionConfigurationForm.php [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ * @package Drupal\encrypt\Form
+ */
+class EncryptionConfigurationForm extends EntityForm {
+
+  /**
+   * KeyManager definition.
+   *
+   * @var \Drupal\key\KeyManager
+   */
+  protected $key_manager;
```
Comment (doylejd) [inline src/Form/EncryptionConfigurationForm.php]: Class variables should be camel case. These should be renamed to $keyManager and $encryptService.

File: src/Form/EncryptionConfigurationForm.php [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+   * Constructs a EncryptionConfigurationForm object.
+   *
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory.
+   * @param \Drupal\Key\KeyManager $key_manager
+   *   The ConditionManager for building the visibility UI.
+   * @param \Drupal\Encrypt\EncryptService $encrypt_service
+   *   The lazy context repository service.
+   */
+  public function __construct(ConfigFactoryInterface $config_factory, KeyManager $key_manager, EncryptService $encrypt_service) {
```
Comment (doylejd) [inline src/Form/EncryptionConfigurationForm.php]: Is $config_factory being used here? If not, can we remove it?

**Comment by nerdstein** [comment]
Issue exists with allowing Key to be "configurable", this means the key would need to be in code and thats a horrible thing for security.

One "idea" to explode - is to decouple they Key "settings" from the Encryption Instances, keep the "Key" global as a setting, make the encryption instances reference that

**Comment by nerdstein** [comment]
Rename to "encryption profile"

**Comment by nerdstein** [comment]
OK this is finally ready to go once the Key module PR gets committed

**Comment by rlhawk** [comment]
On installation, I'm getting a fatal error. 
```
Fatal error: require(): Failed opening required '/var/www/drupal/modules/contrib/encrypt/src/Form/EncryptSettingsForm.php' 
```

**Comment by rlhawk** [comment]
It looks as though this line needs to be changed in the routing file.

`    _form: '\Drupal\encrypt\Form\EncryptSettingsForm'`

**Comment by rlhawk** [comment]
Changing to
`    _form: '\Drupal\encrypt\Form\ProfileForm'`
fixed the error.

**Comment by rlhawk** [comment]
I meant
`    _form: '\Drupal\encrypt\Form\EncryptionProfileForm'`

**Comment by nerdstein** [comment]
Working on the routing changes now

**Comment by rlhawk** [comment]
This architecture looks generally good to me.

**Comment by nerdstein** [comment]
Merging

**Pull request merged**